SatelliteQE/robottelo
# New Bugzilla skipping logic - Replacing Robozilla Issue #7336

Fix #7336 
Replaces PR: #7346 

#### Replaces Robozilla decorators with Pytest Markers and adds new logic for skipping and deselect.

This PR is separated in 5 commits. 
(when merging it can be squased in to one commit, or just rebased keeping 5 different commits)

1. Cleanup of redundant decorators from test cases https://github.com/SatelliteQE/robottelo/pull/7404/commits/23197ae9d3df38786643b33b19b15c3595c694eb
2. Fix of Robottelo Unit Tests for the decorators changes https://github.com/SatelliteQE/robottelo/pull/7404/commits/25e1fb0e8a9d12491ddf673010709cac21cb8074
3. Cleanup of redundant bz functions used within robottelo modules https://github.com/SatelliteQE/robottelo/pull/7404/commits/1e8b9df0725d8dbcda718b00b7f3467e7a20d120
4. Simple .gitignore and Makefile changes https://github.com/SatelliteQE/robottelo/pull/7404/commits/bc35716de207455d14fbf5d54fa56beee34ab582
5. Replaces Robozilla decorators with Pytest markers and adds new logic for `is_open` function https://github.com/SatelliteQE/robottelo/pull/7404/commits/6d4f0fc3de550b851bd2823294ec2acc31acf585

#### The commit n.5 is the most important:
**Replaces Robozilla decorators with Pytest markers and adds new logic for `is_open` function https://github.com/SatelliteQE/robottelo/pull/7404/commits/6d4f0fc3de550b851bd2823294ec2acc31acf585**
Its logic is based on flowcharts described here -> https://hackmd.io/01mIqiG6TOKTrPpTk4jejg?view

At the `files changed` tab reviewers can select which commit to review.

### Cleanups

I checked every BZ using the old `skip_if_bug_open` and also `bz_bug_is_open` function and cleaned up the code when the BZ status was evaluating the function always to True. (in other workds we were doing useless calls to Bz API)

### BZ REST API

> **IMPORTANT!!!**

The new implementation is using `httpx` module to call Bugzilla REST API, it introduces changes on `robottelo.properties` file, that will need to be changed to include `api_key` under `[bugzilla]` section.

So CI configs and also your local configs will need to be updated.

### The new skipping implementation

- Uses httpx to communicate with Robozilla REST API
- Authenticates using `api_key` (that you can generate in your BZ profile page)
- Makes a single call to get all the BZs data at the test collection phase.
- Uses `tenacity.retry` to perform retries if HTTP connection fails.
- Can bypass the api call if `--bz-cache` is passed to pytest then it will use pre-fetched data (useful for local testing).
- Using `pytest tests/foreman --collect-only` will create a new `bz_cache.json` on the root of folder so this file can be used as `--bz-cache` for future runs, and also can be used by the nagger script.
- Has simpler logic based on https://hackmd.io/01mIqiG6TOKTrPpTk4jejg?view

### Usage

#### To skip a test case based on BZ status use:

```python
import pytest

@pytest.mark.skip_if_open('BZ:123456')
def test_that_will_skip_if_bz_is_not_yet_fixed_for_current_system_version(...)
    ...
```

#### To deselect a test case based on BZ status use:

```python
import pytest

@pytest.mark.skip_if_open('BZ:123456')
def test_that_will_be_deselected_from_testing_collection_if_bz_is_wontfix(...)
    # WONTFIX,CANTFIX,DEFERRED makes this test to be ignored by pytest runner
    ...
```

#### To use a `workaround` inside a test case

```python
from robottelo.helpers import is_open

def test_something(...):
    ...
    if is_open('BZ:213456'):
        # this block executes only if BZ is not yet fixed for current system version
```

### The nagger script

There will still be a nagger script, that will notify weekly, component owners to do cleanups of redundant marks, the nagger script will use the `bz_cache.json` generated on every robottelo run.

### TO-DOs

- [x] Send a PR to gitlab config adding new `[bugzilla]`  configs
- [ ] Send PR to robottelo-ci saving the `bz_cache.json` artifact on every build.

### How to test this PR?

- checkout to this branch 
  if you have `hub` so `hub pr checkout 7404`
- install the new dependencies `httpx` and `tenacity`
  `pip install -r requirements.txt`
- Change your `robottelo.properties` adding
   ```ini
    [bugzilla]
    url=https://bugzilla.redhat.com
    api_key=(generate your key on bugzilla profile page)
   ```
- Run `make test-robottelo`
- Run `py.test -s tests/foreman/ --collect-only` 
  No test will run, it will just collect data and generate the `bz_cache.json` take a look at this file :)
- Run `py.test -s tests/foreman/ --collect-only --bz-cache` 
  No test will run, it will just collect data use existing `bz_cache.json`
- Pick some test marked with `@pytest.mark.skip_if_open()` and run it separatelly to see if it works.
